### ISSUE TITLE
Unexpected results using multiple DominatorTree instances

### ISSUE THREAD

**Issue by Waldo1637** [Sable/soot#382]
Unexpected results occur using two instances of DominatorTree, for example one pre-dominator and one post-dominator formed from the same graph:
```java
DominatorTree<N> A = new MHGDominatorTree<>(new MHGDominatorsFinder<>(graph));
DominatorTree<N> B = new MHGDominatorTree<>(new MHGPostDominatorsFinder<>(graph));
```
When calling a function on B with a DominatorNode parameter (i.e. , getParentOf, getChildrenOf, and isImmediateDominatorOf) we expect the result to depend on the structure of tree B regardless of which tree the parameter was from. For example, `B.getParentOf(A.getHead())` is expected to return the parent of `A.getHead()` in graph B however, it returns `null` (since `A.getHead()` has no parent in graph A).

**Comment by ericbodden** [comment]
Hi.

Sorry I don't understand the problem. Can you rephrase, maybe provide a test case?

Cheers,
Eric

> On 01.04.2015, at 19:31, Timothy Hoffman <notifications@github.com> wrote:
> 
> Unexpected results occur using two instances of DominatorTree, for example one pre-dominator and one post-dominator formed from the same graph:
> 
> DominatorTree<N> A = new MHGDominatorTree<>(new MHGDominatorsFinder<>
> (graph));
> 
> DominatorTree<N> B = new MHGDominatorTree<>(new MHGPostDominatorsFinder<>(graph));
> When calling a function on B with a DominatorNode parameter (i.e. , getParentOf, getChildrenOf, and isImmediateDominatorOf) we expect the result to depend on the structure of tree B regardless of which tree the parameter was from. For example, B.getParentOf(A.getHead()) is expected to return the parent of A.getHead() in graph B however, it returns null (since A.getHead() has no parent in graph A).
> 
> —
> Reply to this email directly or view it on GitHub.
> 

--
Prof. Eric Bodden, Ph.D., http://sse.ec-spride.de/ http://bodden.de/
Head of Secure Software Engineering at Fraunhofer SIT, TU Darmstadt and EC SPRIDE
Tel: +49 6151 16-75422    Fax: +49 6151 869-127
Room B5.11, Fraunhofer SIT, Rheinstraße 75, 64295 Darmstadt

**Comment by StevenArzt** [comment]
I don't think this should be a problem in the graph. You are taking a node from one graph and calling a function on a totally different graph with that node. Even if both nodes are sued to model the same element outside of the tree, they remain distinct tree nodes.  Therefore, in my opinion, any matching between different trees is in the responsibility of the user. The tree in itself is correct: It works with its own nodes.

**Comment by Waldo1637** [comment]
I agree that the implementation is correct when using nodes from the tree itself. I also realize that simultaneously using DominatorTree instances built from the same graph is an uncommon use case. However, I made the suggestion because I think it's counterintuitive to call a function on a tree and get results that are not consistent with the structure of that tree itself.
The solution I proposed in the associated pull request is quite simple. Maybe an alternative solution would be to throw an error if the "dode" passed as an argument does not belong to the current tree. I just believe that the current behavior may not be the best solution. Thanks for your consideration of this issue.

**Comment by ericbodden** [comment]
@Waldo1637 I would suggest to maximally throw an exception in such cases. Indeed this use case should be super uncommon and I think it really depends on the use case what behavior users would like to see when issuing such calls. Users can always create a subclass with their desired behavior.

**Comment by StevenArzt** [comment]
I think a return value of null is perfectly fine, so I'll close the issue as we have agreed that we won't merge the proposed code changes.
